xbmc/cores/VideoPlayer/DVDCodecs/Video/VAAPI.cpp

```diff
@@ -1947,85 +1942,125 @@ void COutput::InitCycle()
     if (!m_config.processInfo->Supports(method))
       method = VS_INTERLACEMETHOD_VAAPI_BOB;
 
-    if (m_pp && !m_pp->UpdateDeintMethod(method))
-    {
-      std::shared_ptr<CPostproc> pp(m_pp);
-      m_discardedPostprocs.push_back(pp);
-      m_pp->Discard(this, &COutput::ReadyForDisposal);
-      m_pp = nullptr;
-      m_config.processInfo->SetVideoDeintMethod("unknown");
-    }
-    if (!m_pp)
+    if (m_currentDiMethod != method)
     {
-      if (method == VS_INTERLACEMETHOD_DEINTERLACE ||
-          method == VS_INTERLACEMETHOD_RENDER_BOB)
+      if (m_pp && !m_pp->UpdateDeintMethod(method))
       {
-        m_pp = new CFFmpegPostproc();
-        m_config.stats->SetVpp(false);
-      }
-      else
-      {
-        m_pp = new CVppPostproc();
-        m_config.stats->SetVpp(true);
+        CLog::Log(LOGDEBUG, LOGVIDEO, "VAAPI output: Current postproc does not want new deinterlace mode, removing");
+        std::shared_ptr<CPostproc> pp(m_pp);
+        m_discardedPostprocs.push_back(pp);
+        m_pp->Discard(this, &COutput::ReadyForDisposal);
+        m_pp = nullptr;
+        m_config.processInfo->SetVideoDeintMethod("unknown");
       }
-      if (m_pp->PreInit(m_config))
+      if (!m_pp)
       {
-        m_pp->Init(method);
-
-        if (method == VS_INTERLACEMETHOD_DEINTERLACE)
-          m_config.processInfo->SetVideoDeintMethod("yadif");
-        else if (method == VS_INTERLACEMETHOD_RENDER_BOB)
-          m_config.processInfo->SetVideoDeintMethod("render-bob");
-        else if (method == VS_INTERLACEMETHOD_VAAPI_BOB)
-          m_config.processInfo->SetVideoDeintMethod("vaapi-bob");
-        else if (method == VS_INTERLACEMETHOD_VAAPI_MADI)
-          m_config.processInfo->SetVideoDeintMethod("vaapi-madi");
-        else if (method == VS_INTERLACEMETHOD_VAAPI_MACI)
-          m_config.processInfo->SetVideoDeintMethod("vaapi-mcdi");
-      }
-      else
-      {
-        delete m_pp;
-        m_pp = NULL;
+        if (method == VS_INTERLACEMETHOD_DEINTERLACE ||
+            method == VS_INTERLACEMETHOD_RENDER_BOB)
+        {
+          CLog::Log(LOGDEBUG, LOGVIDEO, "VAAPI output: Initializing ffmpeg postproc");
+          m_pp = new CFFmpegPostproc();
+          m_config.stats->SetVpp(false);
+        }
+        else
+        {
+          CLog::Log(LOGDEBUG, LOGVIDEO, "VAAPI output: Initializing vaapi postproc");
+          m_pp = new CVppPostproc();
+          m_config.stats->SetVpp(true);
+        }
+        if (m_pp->PreInit(m_config))
+        {
+          m_pp->Init(method);
+        }
+        else
+        {
+          CLog::Log(LOGERROR, "VAAPI output: Postproc preinit failed");
+          delete m_pp;
+          m_pp = NULL;
+        }
       }
+
+      if (method == VS_INTERLACEMETHOD_DEINTERLACE)
+        m_config.processInfo->SetVideoDeintMethod("yadif");
```

Comment:
Setting deinit method can be done by Cpostproc. It has access to config. Then resurrecting m_currentDiMethod is not needed